### ISSUE TITLE
SUCCESSOR_TAB_CONTROL_IN_TREE should skip hidden tabs

### ISSUE THREAD

**Issue by pintassilgo** [piroor/treestyletab#2453]
**tl;dr** a hidden tab (i.e. a tab from another group) should never be chosen to be focused after closing a tab.

---

STR

1. Install TST;
2. Install Simple Tab Groups (any tab groups addon should give the same result, this one is a recommended extension and the most popular of its class);
3. TST Options > select `Fix up tree structure with visible tabs automatically (*Recommended if you use another addon to switch tab groups)`;
4. Put open tabs in groups (this step is needed because STG doesn't create any group when installed). It can be only one tab in one group. Close other tabs, select the one left, rightclick it then `Move tab to group > Create a new group`.
5. Focus on the tab you moved. Open a child tab. The tree in the group must look like this:
```
Tab 1
-> Tab 2
```
6. Select Tab 1 and move it to another group. You should notice that Tab 2 will be detached as a root tab. But not a real root, as you will soon realize.
7. Select Tab 1 in the other group you put it in.
8. Open a new independent tab (let's call it Tab 3) just to ensure that the group will continue to exist, because you will close Tab 1.
9. Focus Tab 1 and close it.

Expected result: Tab 3 should be focused, since it's the only visible tab i.e. the remaining tab in current group.

Actual result: Tab 2 is focused, so the view changes to the other group.

I think a hidden tab should never be chosen to be focused after closing a tab (unless the closed tab was the last visible one).

Note.: just to avoid any misunderstanding, I'm talking about hidden tab, not collapsed tab.

**tl;dr** a hidden tab should never be chosen to be focused after closing a tab.

**Comment by piroor** [comment]
This doesn't happen after the commit b06a2061951203ee700e6414cbdeb3816f41d9a3. Both issues look to be caused by the same reason. I think this can be treated as fixed.

**Comment by pintassilgo** [comment]
This issue is not fixed yet. I'll try to get a consistent steps to reproduce.

**Comment by pintassilgo** [comment]
STR

**1.** Install TST 3.3.2.9938;
**2.** TST Options > select `Fix up tree structure with visible tabs automatically (*Recommended if you use another addon to switch tab groups)`;
**3.** Install [Tiled Tab Groups 0.18.12](https://addons.mozilla.org/en-US/firefox/addon/tiled-tab-groups/)
**4.** Open a new tab, rightclick it, `Move Tab to group > Move Tab to New Group`. The view will change to this new group.
**5.** Go back to the other group (click Tiled Tab Groups button, then Group 0);
**6.** Open example.com, then open the only link on the page in a child tab (I did it by middleclicking);
**7.** Open a new tab. Now your tabbar should look like this:

* about:addons
* example.com (expanded tree)
   * IANA
* New Tab

And another New Tab hidden in the other group.

**8.** Focus on IANA tab and move it to Group 1. The view will change to that group, with IANA being the second tab in the bar, next to "New Tab".

**9.** Close current tab (IANA).

You should remain in Group 1, with the only tab left being focused (New Tab). Instead, the view switches back to Group 0.

**Comment by piroor** [comment]
Thanks a  lot! Finally I've realized that TST sometimes didn't consider visibility of tabs when it finds a successor for the current tab. The case you reported should be fixed with recent commits.
